diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2469fc4..ef0a32d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -903,3 +903,5 @@ include(CTest)
 add_subdirectory(lib)
 add_subdirectory(tests)
 add_subdirectory(doc)
+
+configure_file(config.py.in ${CMAKE_INSTALL_PREFIX}/config.py)
diff --git a/config.py.in b/config.py.in
new file mode 100644
index 0000000..2f4bfb5
--- /dev/null
+++ b/config.py.in
@@ -0,0 +1,6 @@
+"""
+This file is automitically generated during the setup.
+Please do not edit config.py, but edit config.py.in in the patches directory.
+"""
+
+QUDA_MPI = "${QUDA_MPI}" == "ON"
